### ISSUE TITLE
binary incompatibility

### ISSUE THREAD

**Issue by svlandeg** [huggingface/neuralcoref#158]
I'm using the current spaCy from the master branch, and getting this error: 
> RuntimeWarning: spacy.tokens.span.Span size changed, may indicate binary incompatibility. Expected 72 from C header, got 80 from PyObject

I'm assuming this happens because `span.pxd` has changed after the 2.1 release: https://github.com/explosion/spaCy/commits/master/spacy/tokens/span.pxd

I tried reinstalling with 
> pip install neuralcoref --no-binary neuralcoref

But the warning remains and the program crashes when I run `nlp(doc)`:

> Process finished with exit code -1073741819 (0xC0000005)

Any idea on how to fix this? I'm compiling spaCy from sources too, so I was hoping not to have to do the same for neuralcoref ...

**Comment by ndeku** [comment]
I had the same issue. I downgraded Spacy from 2.1.4 to 2.1.0 and numpy to 1.14.5 to resolve.

This Spacy issue seems relevant (and still unresolved?): https://github.com/explosion/spaCy/issues/2798

**Comment by Vergir** [comment]
Had same issue, downgrading spacy to 2.1.0 worked for me as well.

**Comment by svlandeg** [comment]
That's not an option for me unfortunately, I need certain new features in spaCy from after the 2.1 release...

**Comment by cheevahagadog** [comment]
I too had to uninstall neuralcoref and re-install without binaries. Then I downgraded spaCy from 2.1.4 to 2.1.0 but I didn't have to change my numpy (still at 1.15.0). Seems to be working great now.

**Comment by Gunjitbedi** [comment]
Yes, downgrading to 2.1.0 does seems to work

**Comment by BramVanroy** [comment]
At first I was afraid this was a Windows related issue. Glad that it is "only" a version issue.

**Comment by BramVanroy** [comment]
@svlandeg Not sure why you closed this? I guess it can only be closed after pulling https://github.com/huggingface/neuralcoref/pull/186

**Comment by svlandeg** [comment]
Pinning spaCy to `2.1.0` will prevent this issue for most users, but does not support my use-case where I actually want to use a newer version of spaCy (compiling the current master branch from source). I guess there's always going to be some lag between spaCy changes and a new neuralcoref version, right? So that's why I closed it.

**Comment by BramVanroy** [comment]
If you can compile from source, and it works, then I'm a bit confused why the pip distribution is lagging behind? If the current source works, pinning to 2.1.0 is not necessary indeed, but then I would expect a distributed version update too. I don't mean that in a "I WANT THIS NOW" way, but it is kind of odd that such an important issue is not pushed to package managers as soon as possible - because the current version simply does not work as-is, since the newest version of spaCy will be installed and then neuralcoref will fail. Is there a specific reason why the changes have not been distributed? Apart from that, closing an issue without an explanation to the users in this topic causes more confusion, I think so it's good that we have this discussion now to clarify it.

**Comment by svlandeg** [comment]
I think there's some miscommunication here... I think your PR of pinning spaCy to 2.1.0 makes a lot of sense and will prevent this issue from popping up for other users. I'm happy to reopen this issue if you think that's necessary.

I initially closed this issue because a solution was provided (downgrading) that seemed to satisfy all other users, and because I started realising that a straightforward solution for my original question is probably not possible.

**Comment by thomwolf** [comment]
Yes, this is

**Comment by svlandeg** [comment]
@thomwolf: So just to summarize, I think there are two main issues discussed in this thread:

- Using `neuralcoref` out-of-the-box will install the latest `spaCy` version because of `spacy>=2.1.0` in the `requirements.txt`, but the latest version of spaCy has changes in some of its core Cython files (such as `span.pxd`) so this raises incompatibility issues. For most users, downgrading to spaCy 2.1.0 will be fine for now, so the PR by @BramVanroy makes sense to address this issue: https://github.com/huggingface/neuralcoref/pull/186

- When you do want to use a newer spaCy version > 2.1.0, like I do, a `binary incompatibility` RuntimeWarning is thrown by neuralcoref. According to the [docs](https://github.com/huggingface/neuralcoref#install-neuralcoref-with-pip), this should be fixed by installing with `pip install neuralcoref --no-binary neuralcoref` instead, but that doesn't seem to fix it for me. The docs do mention this as a fix for a warning around `spacy.strings.StringStore size changed` while I might add that the current incompatibility is caused by `spacy.tokens.span.Span`.

**Comment by colingoldberg** [comment]
I don't want to add noise to this thread, but having got the same error message (see below), I want to get a notification when this is fixed. I have provided an error log (if this helps).

[hs_err_pid2470.log](https://github.com/huggingface/neuralcoref/files/3427277/hs_err_pid2470.log)


Steps:
- Error message: `/Users/.../anaconda3/lib/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: spacy.tokens.span.Span size changed, may indicate binary incompatibility. Expected 72 from C header, got 80 from PyObject
  return f(*args, **kwds)
Segmentation fault: 11`
- In case the following Java error info is useful, I have included it here (got it while running under Visual Studio Code)
`
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x0000000138c31ff3, pid=2470, tid=775
#
# JRE version: Java(TM) SE Runtime Environment (11.0.1+13) (build 11.0.1+13-LTS)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (11.0.1+13-LTS, mixed mode, tiered, compressed oops, g1 gc, bsd-amd64)
# Problematic frame:
# C  [neuralcoref.cpython-36m-darwin.so+0x27ff3]  _ZL60__pyx_pf_11neuralcoref_11neuralcoref_11NeuralCoref_13predictP49__pyx_obj_11neuralcoref_11neuralcoref_NeuralCorefP7_objectfiiS2_i+0x1523
#
# No core dump will be written. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/.../Development/nlp/spaCy/tests/hs_err_pid2470.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#
Abort trap: 6
`

- As per instructions re binary incompatibility, I uninstalled and re-installed with "--no-binary neuralcoref", but got same result

Macosx 10.14.6
* **spaCy version:** 2.1.4
* **Platform:** Darwin-18.6.0-x86_64-i386-64bit
* **Python version:** 3.6.7

Colin Goldberg

**Comment by BramVanroy** [comment]
How is Java involved in this?

**Comment by colingoldberg** [comment]
Probably because I was running inside Visual Studio Code. if not, I do not know.

**Comment by thomwolf** [comment]
@svlandeg can you confirm that if you build neuralcoref's current master branche from source while having a spacy > 2.1.0 it works?

**Comment by BramVanroy** [comment]
I installed the current commit (https://github.com/huggingface/neuralcoref/commit/3d12574a9e9474060d8bfbbe3fa8751e9738a3fa) successfully with Python 3.7.3, spaCy 2.1.6.

**Comment by svlandeg** [comment]
> 
> 
> @svlandeg can you confirm that if you build neuralcoref's current master branche from source while having a spacy > 2.1.0 it works?

Yep, when I build neuralcoref's master branch from source, in combination with the current spaCy master branch from source, it does work.

**Comment by jonathankingfc** [comment]
@svlandeg I am having trouble following your steps. I am working inside of a Docker container. I am currently just using SpaCy 2.1.3 and Neuralcoref 4.0, but I would like to upgrade to 2.1.7 due to the improved NER scorer. I followed the instructions to build both from source inside of my Dockerfile, but it seems that either Neuralcoref or SpaCy is removed from the Python module path. I think it has something to do with them being codependencies. Any help would be greatly appreciated!

**Comment by svlandeg** [comment]
Hi @jonathankingfc : not sure how to help you without having access to your console ;-) I just created a new environment and installed spaCy from source, and then did the same with neuralcoref. The current `requirements.txt` for neuralcoref reads `spacy>=2.1.0` so if you've installed spaCy 2.1.7 from source, that shouldn't conflict.

**Comment by honnibal** [comment]
> According to the docs, this should be fixed by installing with pip install neuralcoref --no-binary neuralcoref instead, but that doesn't seem to fix it for me. 

The `--no-binary` flag will tell pip to install the `sdist`, rather than the binary wheel. The problem is that the `sdist` of a Cython package like neuralcoref isn't *really* the source: it already has some compilation --- the Python-to-Cython processing stage.

The problem is that the sdist includes .cpp files that were generated using the .pxd files of a particular spaCy version. These are basically headers that get copied into the .cpp source when you run Cython. Those headers specify a different size for some objects, e.g. Span.

So what's needed is a new Neuralcoref release that pins against the new spaCy. As soon as the new sdist is generated, everything should work correctly --- because Cython just needs to be rerun with the new spaCy. That's why building from master works, as @svlandeg reports above.

**Comment by MRocholl** [comment]
That fixes it for me.

**Comment by sent1nel** [comment]
Does not fix the issue for me. Going to try to build from sources...

**Comment by fhamborg** [comment]
getting the same error, installed spacy and neuralcoref as follows:
```
conda install -c conda-forge spacy
python -m spacy download en_core_web_lg
pip install neuralcoref
```

i also tried the "fix description" from neuralcoref's readme - error is identical though:
```
pip uninstall neuralcoref
pip install neuralcoref --no-binary neuralcoref
```

this is the error i'm getting:
``` 
[path]/python3.6/importlib/_bootstrap.py:219: RuntimeWarning: spacy.tokens.span.Span size changed, may indicate binary incompatibility. Expected 72 from C header, got 80 from PyObject
  return f(*args, **kwds)
```

**Comment by sent1nel** [comment]
@fhamborg did you try compiling manually neuralcoref?

```
git clone https://github.com/huggingface/neuralcoref.git
cd neuralcoref
pip install -r requirements.txt
pip install -e .
```

**Comment by fhamborg** [comment]
yes, i tried that, too, but got dependency issues with other packages in the realm of pandas, numpy and the like. i managed to circumvent the issue by not using conda on python 3.6 but plain pip on python 3.5 for spacy. what works for me (just posting this here in case others have a similar issue and use maybe similar data science/ML packages - so it's partially off topic):
```
conda create -y --name cope python=3.5
conda activate cope
pip install spacy==2.1.0
python -m spacy download en_core_web_sm
python -m spacy download en_core_web_lg
pip install neuralcoref
pip install torch
pip install pytorch-transformers
pip install pandas
```

**Comment by svlandeg** [comment]
Closing this as a solution is provided (either downgrade `spaCy `or install both `spaCy` and `neuralcoref` from source). The underlying issue of having up-to-date releases with the latest version of `spaCy` is merged with Issue #197

**Comment by fhamborg** [comment]
IMO that's neither a fix nor solution but just a workaround. For instance, if one needs to use functionality provided just in spacy's more recent versions (>2.1.0), there's currently not way to use neuralcoref, too. Installing both neuralcoref and spacy from source did not help me, the only workaround was to use spacy on 2.1.0

**Comment by svlandeg** [comment]
Hi @fhamborg : that's odd, because I can definitely install both from source and work with spaCy 2.1.6 and neuralcoref 4 in the same env, and others have confirmed the same above. If you're still running into issues, feel free to open a new issue and post a log of your console commands & error trace.

I agree with you though that this workaround is not ideal. We will be looking into a more streamlined way to release new version as the underlying Cython structures change in spaCy, to keep the two more in-sync (cf Issue #197).

**Comment by ThomasMuhr** [comment]
dll hell is nothing compared to this versions chaos!-/
